slkspec/core.py

```diff
@@ -1,129 +1,355 @@
+        self._file_queue = _file_queue
+        print(self._file)
+        with _lock:
+            if not Path(self._file).exists() or override:
+                self._file_queue.put((self._url, str(Path(self._file).parent)))
+            elif Path(self._file).exists():
+                if self.touch:
+                    Path(self._file).touch()
+                self._file_obj = open(self._file, mode, **kwargs)
+
+    @property
+    def name(self) -> str:
+        """Get the file for the SLKFile object."""
+        if self._file_obj is not None:
+            return self._file
+        return self._url
+
+    def _retrieve_items(self, retrieve_files: list[tuple[str, str]]) -> None:
+        """Get items from the tape archive."""
+
+        retrieval_requests: Dict[Path, List[str]] = defaultdict(list)
+        logger.critical("Retrieving %i items from tape", len(retrieve_files))
+        for inp_file, out_dir in retrieve_files:
+            retrieval_requests[Path(out_dir)].append(inp_file)
+        for output_dir, inp_files in retrieval_requests.items():
+            output_dir.mkdir(parents=True, exist_ok=True, mode=self.file_permissions)
+            logger.critical("Creating slk query for %i files", len(inp_files))
+            search_str = pyslk.slk_search(pyslk.slk_gen_file_query(inp_files))
+            search_id_re = re.search("Search ID: [0-9]*", search_str)
+            if search_id_re is None:
+                raise ValueError("No files found in archive.")
+            search_id = int(search_id_re.group(0)[11:])
+            logger.critical("Retrieving files for search id: %i", search_id)
+            pyslk.slk_retrieve(search_id, str(output_dir))
+            logger.critical("Adjusting file permissions")
+            for out_file in map(Path, inp_files):
+                out_file.chmod(self.file_permissions)
+
+    def _cache_files(self) -> None:
+        with self._lock:
+            items = []
+            if self._file_queue.qsize() > 0:
+                self._file_queue.put(("finish", "finish"))
+                for _ in range(self._file_queue.qsize() - 1):
+                    items.append(self._file_queue.get())
+                    self._file_queue.task_done()
+                self._retrieve_items(items)
+                _ = self._file_queue.get()
+                self._file_queue.task_done()
+        self._file_queue.join()
+        self._file_obj = open(self._file, self.mode, **self.kwargs)
+
+    def __fspath__(self) -> str:
+        if self._file_obj is None:
+            self._cache_files()
+        return self.name
+
+    def tell(self) -> int:
+        if self._file_obj is not None:
+            return self._file_obj.tell()
+        self._cache_files()
+        return self._file_obj.tell()  # type: ignore
+
+    def seek(self, target: int) -> int:  # type: ignore
+        if self._file_obj is None:
+            return self._file_obj.seek(target)  # type: ignore
+        self._cache_files()
```

Comment:
> I would suggest that we deal with getting the preffs to work later. Unless you see what the issue is.

@antarcticrainforest, I think this is the issue here. In case `self._file_obj is None` it won't have a `.seek` ;)

Reply:
Cheez, of course!!! Thanks.